json/test_yajl.c

```diff
@@ -0,0 +1,184 @@
+	context_t* ctx = ctx_v;
+
+	switch (ctx->item) {
+		case 'x': ctx->c.x += value; break;
+		case 'y': ctx->c.y += value; break;
+		case 'z': ctx->c.z += value; break;
+		default:;
+	}
+
+	return 1;
+}        
+
+int cb_start_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth++;
+
+	if (ctx->depth == 2 && ctx->in_coord) {
+		ctx->len++;
+	}
+
+	return 1;
+}
+
+int cb_map_key(void* ctx_v, const unsigned char* key, size_t length) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1) {
+		if (strncmp("coordinates", (const char*)key, length) == 0) {
+			ctx->in_coord = 1;
+		} else {
+			ctx->in_coord = 0;
+		}
+	} else if (ctx->depth == 2 && ctx->in_coord) {
+		if (strncmp("x", (const char*)key, length) == 0) {
+			ctx->item = 'x';
+		} else if (strncmp("y", (const char*)key, length) == 0) {
+			ctx->item = 'y';
+		} else if (strncmp("z", (const char*)key, length) == 0) {
+			ctx->item = 'z';
+		} else {
+			ctx->item = 0;
+		}
+	}
+
+	return 1;
+}
+
+int cb_end_map(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	ctx->depth--;
+
+	return 1;
+}
+
+int cb_end_array(void* ctx_v) {
+	context_t* ctx = ctx_v;
+
+	if (ctx->depth == 1 && ctx->in_coord) {
+		ctx->c.x /= ctx->len;
+		ctx->c.y /= ctx->len;
+		ctx->c.z /= ctx->len;
+	}
+
+	return 1;
+}
+
+static yajl_callbacks ycallbacks = {
+	NULL, /* null */
+	NULL, /* boolean */
+	NULL, /* integer */
+	&cb_double,
+	NULL, /* number */
+	NULL, /* string */
+	&cb_start_map,
+	&cb_map_key,
+	&cb_end_map,
+	NULL, /* start_array */
+	&cb_end_array
+};
+
+#define EMPTY_CTX ((context_t){0, 0, 0, 0, {0.0, 0.0, 0.0}})
+
+int main() {
+	const char test_buffer_1[] = "{\"coordinates\":[{\"x\":2.0,\"y\":0.5,\"z\":0.25}]}";
+	const char test_buffer_2[] = "{\"coordinates\":[{\"y\":0.5,\"x\":2.0,\"z\":0.25}]}";
+
+	context_t ctx = EMPTY_CTX;
+	yajl_handle yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+
+	yajl_parse(yajl, (const unsigned char*)test_buffer_1, sizeof(test_buffer_1) / sizeof(test_buffer_1[0]));
+	yajl_complete_parse(yajl);
+
+	if (!(ctx.c.x == 2.0 && ctx.c.y == 0.5 && ctx.c.z == 0.25)) {
+		fprintf(stderr, "{%lf, %lf, %lf} != {2.0, 0.5, 0.25}\n", ctx.c.x, ctx.c.y, ctx.c.z);
+		exit(EXIT_FAILURE);
+	}
+
+	yajl_free(yajl);
+
+	ctx = EMPTY_CTX;
+	yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+	yajl_parse(yajl, (const unsigned char*)test_buffer_2, sizeof(test_buffer_2) / sizeof(test_buffer_2[0]));
+	yajl_complete_parse(yajl);
+
+	if (!(ctx.c.x == 2.0 && ctx.c.y == 0.5 && ctx.c.z == 0.25)) {
+		fprintf(stderr, "{%lf, %lf, %lf} != {2.0, 0.5, 0.25}\n", ctx.c.x, ctx.c.y, ctx.c.z);
+		exit(EXIT_FAILURE);
+	}
+
+	yajl_free(yajl);
+
+	int fd = open("/tmp/1.json", O_RDONLY);
+	int ret = 0;
+	unsigned char buf[4096];
+
+	if (fd < 0) {
+		fprintf(stderr, "Cannot open /tmp/1.json: %s\n", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	ctx = EMPTY_CTX;
+	yajl = yajl_alloc(&ycallbacks, NULL, &ctx);
+
+	notify_with_pid("C/" COMPILER " (yajl)");
+
+	while ((ret = read(fd, buf, sizeof(buf)/sizeof(buf[0])))) {
```

Comment:
All tests parse already loaded buffer in the memory. Please change the logic to load the content first, and then parse it.

Reply:
Hello.

While I agree with other comments, I have a question here. `yajl` is streaming parsing library, so if we load all the data into the memory we (unnecessarily) increase the test memory consumption. Should I make two flavors of this test? With the loaded buffer parsing and with streaming reading.

Reply:
The increased memory consumption is expected. Streaming reading is out of scope for those particular tests - they've asked for that, but that would make the results table way too big.

Plus the streaming results could be confusing. For example, one of the most efficient way to do "streaming" reading is using `mmap`, but it's way too unrealistic for the majority of use-cases (as usually JSON is used to transfer data). Using network streams is closer to the production use-cases, but makes tests way too complicated. So, the common ground now is just parse buffers.